File: building-a-wallet/python/step4_send_sol.py
import asyncio
import httpx
from solana.rpc.api import Client
from solana.rpc.types import TxOpts
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.system_program import TransferParams, transfer
from solders.transaction import Transaction as SolTransaction
from solders.message import Message
from solders.compute_budget import set_compute_unit_price, set_compute_unit_limit

ENDPOINT = "https://your-endpoint.rpcpool.com"
TOKEN = "your-token"

async def get_priority_fee(accounts: list[str], percentile: int) -> int:
    async with httpx.AsyncClient() as http:
        resp = await http.post(
            f"{ENDPOINT}/{TOKEN}",
            json={
                "jsonrpc": "2.0", "id": 1,
                "method": "getRecentPrioritizationFees",
                "params": [accounts, {"percentile": percentile}],
            },
        )
        fees = resp.json()["result"]
        fees.sort(key=lambda x: x["slot"], reverse=True)
        return fees[0]["prioritizationFee"]

def send_sol(sender: Keypair, recipient_address: str, lamports: int, priority_fee: int) -> str:
    client = Client(f"{ENDPOINT}/{TOKEN}")
    recipient = Pubkey.from_string(recipient_address)

    blockhash = client.get_latest_blockhash().value.blockhash

    instructions = [
        set_compute_unit_limit(200_000),
        set_compute_unit_price(priority_fee),
        transfer(TransferParams(
            from_pubkey=sender.pubkey(),
            to_pubkey=recipient,
            lamports=lamports,
        )),
    ]

    msg = Message.new_with_blockhash(instructions, sender.pubkey(), blockhash)
    tx = SolTransaction([sender], msg, blockhash)

    response = client.send_raw_transaction(
        bytes(tx),
        opts=TxOpts(skip_preflight=True, max_retries=0),
    )
    print(f"Sent: {response.value}")
    return str(response.value)

# Example usage — replace with a real funded keypair and recipient
# async def main():
#     sender = Keypair()  # replace with real keypair
#     priority_fee = await get_priority_fee([str(sender.pubkey())], 9000)
#     send_sol(sender, "RecipientAddressHere", 1_000_000, priority_fee)
# asyncio.run(main())

File: building-a-wallet/typescript/step4-send-sol.ts
import {
  Connection,
  PublicKey,
  SystemProgram,
  Transaction,
  ComputeBudgetProgram,
  Keypair,
  LAMPORTS_PER_SOL,
} from "@solana/web3.js";

const ENDPOINT = "https://your-endpoint.rpcpool.com";
const TOKEN = "your-token";

async function getPriorityFee(accounts: string[], percentile: number): Promise<number> {
  const response = await fetch(`${ENDPOINT}/${TOKEN}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      jsonrpc: "2.0",
      id: 1,
      method: "getRecentPrioritizationFees",
      params: [accounts, { percentile }],
    }),
  });
  const { result } = await response.json();
  const fees: { slot: number; prioritizationFee: number }[] = result;
  return fees.sort((a, b) => b.slot - a.slot)[0].prioritizationFee;
}

async function sendSol(
  sender: Keypair,
  recipientAddress: string,
  amountSol: number
): Promise<string> {
  const connection = new Connection(`${ENDPOINT}/${TOKEN}`);
  const recipient = new PublicKey(recipientAddress);
  const lamports = Math.round(amountSol * LAMPORTS_PER_SOL);

  const priorityFee = await getPriorityFee([sender.publicKey.toBase58()], 9000);

  const { blockhash } = await connection.getLatestBlockhash();
  const tx = new Transaction();
  tx.recentBlockhash = blockhash;
  tx.feePayer = sender.publicKey;

  tx.add(
    ComputeBudgetProgram.setComputeUnitLimit({ units: 200_000 }),
    ComputeBudgetProgram.setComputeUnitPrice({ microLamports: priorityFee }),
    SystemProgram.transfer({
      fromPubkey: sender.publicKey,
      toPubkey: recipient,
      lamports,
    })
  );

  tx.sign(sender);

  // Send via Jet — skipPreflight + maxRetries: 0 for direct leader delivery
  const sendResponse = await fetch(`${ENDPOINT}/${TOKEN}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      jsonrpc: "2.0",
      id: 1,
      method: "sendTransaction",
      params: [
        tx.serialize().toString("base64"),
        { encoding: "base64", skipPreflight: true, maxRetries: 0 },
      ],
    }),
  });

  const { result: signature } = await sendResponse.json();
  console.log("Sent:", signature);
  return signature;
}

// Example usage — replace with a real funded keypair and recipient
// const sender = Keypair.fromSecretKey(Uint8Array.from([...]));
// sendSol(sender, "RecipientAddressHere", 0.001).catch(console.error);
